pkg/ds/farm_test.go

```diff
@@ -777,28 +781,176 @@ func TestMultipleFarms(t *testing.T) {
+			if firstFarm.children[anotherDSData.ID].ds.IsDisabled() {
+				return util.Errorf("Expected daemon set to be enabled in only one farm")
+			}
+			if _, ok := secondFarm.children[anotherDSData.ID]; ok {
+				return util.Errorf("Expected daemon set to be held by only one farm")
+			}
+		} else if _, ok := secondFarm.children[anotherDSData.ID]; ok {
+			if secondFarm.children[anotherDSData.ID].ds.IsDisabled() {
+				return util.Errorf("Expected daemon set to be enabled in only one farm")
+			}
+			if _, ok := firstFarm.children[anotherDSData.ID]; ok {
+				return util.Errorf("Expected daemon set to be held by only one farm")
+			}
+		} else {
+			return util.Errorf("Expected daemon set to be enabled in only one farm")
 		}
-	} else {
-		t.Fatalf("Expected daemon set to be disabled in only one farm")
+		return nil
 	}
+	err = waitForCondition(condition)
+	Assert(t).IsNil(err, "Error enabling daemon set!")
 
 	// Verify node2 is unscheduled
-	labeled, err = applicator.GetLabels(labels.POD, "node2/testPod")
-	Assert(t).IsNil(err, "Expected no error getting labels")
-	Assert(t).IsFalse(labeled.Labels.Has(DSIDLabel), "Expected pod not to have a dsID label")
+	labeled, err = waitForPodLabel(applicator, false, "node2/testPod")
+	Assert(t).IsNil(err, "Expected pod not to have a dsID label")
+}
+
+func waitForPodLabel(applicator labels.Applicator, hasDSIDLabel bool, podPath string) (labels.Labeled, error) {
+	var labeled labels.Labeled
+	var err error
+
+	condition := func() error {
+		var val bool
+		labeled, err = applicator.GetLabels(labels.POD, podPath)
+		if err == nil {
+			val = labeled.Labels.Has(DSIDLabel)
+			if hasDSIDLabel == val {
+				return nil
+			}
+		}
+		return util.Errorf("Expected hasIDLabel to be '%v', got '%v'", hasDSIDLabel, val)
+	}
+
+	err = waitForCondition(condition)
+	return labeled, err
+}
+
+func waitForDisabled(
+	dsf *Farm,
+	dsStore *dsstoretest.FakeDSStore,
+	dsID ds_fields.ID,
+	isDisabled bool,
+) error {
+	var newDS ds_fields.DaemonSet
+	var err error
+
+	condition := func() error {
+		newDS, _, err = dsStore.Get(dsID)
+		if err != nil {
+			return util.Errorf("Expected no error getting daemon set")
+		}
+		if newDS.Disabled != isDisabled {
+			return util.Errorf("Unexpected disabled value. Expected '%v', got '%v'",
+				newDS.Disabled,
+				isDisabled,
+			)
+		}
+		if _, ok := dsf.children[newDS.ID]; !ok {
+			return util.Errorf("Expected farm to create child daemon set with id '%v'", dsID)
+		}
+		if dsf.children[newDS.ID].ds.IsDisabled() != isDisabled {
+			return util.Errorf("Unexpected disabled value in farm. Expected '%v', got '%v'",
+				dsf.children[newDS.ID].ds.IsDisabled(),
+				isDisabled,
+			)
+		}
+		return nil
+	}
+
+	err = waitForCondition(condition)
+	return err
+}
+
+func waitForCreate(dsf *Farm, dsID ds_fields.ID) error {
+	condition := func() error {
+		if _, ok := dsf.children[dsID]; !ok {
```

Comment:
this function (and the others) might be clearer when flipped, e.g. `if ok { return nil }`